src/basedtyping/__init__.py

```diff
@@ -591,6 +587,116 @@ def deco(fn: Callable[[], None]) -> Callable[[], None]: ...
         @deco
         def foo(): ...
     """
-    if not isinstance(fn, FunctionType):  # type: ignore[redundant-expr]
+    if not isinstance(fn, FunctionType):
         raise TypeError(f"{fn} is not a FunctionType")
-    return fn  # type: ignore[unreachable]
+    # https://github.com/KotlinIsland/basedmypy/issues/745
+    return cast("FunctionType[P, T]", fn)
+
+
+class ForwardRef(typing.ForwardRef, _root=True):  # type: ignore[call-arg,misc]
+    """
+    Like `typing.ForwardRef`, but lets older Python versions use newer typing features.
+    Specifically, when evaluated, this transforms `X | Y` into `typing.Union[X, Y]`
+    and `list[X]` into `typing.List[X]` etc. (for all the types made generic in PEP 585)
+    if the original syntax is not supported in the current Python version.
+    """
+
+    # older typing.ForwardRef doesn't have this
+    __slots__ = ["__forward_module__", "__forward_is_class__"]
+
+    def __init__(self, arg: str, *, is_argument=True, module: object = None, is_class=False):
```

Comment:
no idea, typeshed has it as `Any | None`...

Reply:
I wouldn't even know what `Any | None` is supposed to be, because the [typing spec](https://typing.readthedocs.io/en/latest/spec/concepts.html#union-types) defines it as

> The union `T | Any` is not reducible to a simpler form. It represents an unknown static type with lower bound `T`. That is, it represents an unknown set of objects which may be as large as object, or as small as `T`, but no smaller.

This sounds to me like they're talking about `T | Any = {∀x | T <: x <: object}`, which is somewhat similar to the set difference of `object - T`.
But that's such a ridiculously absurd definition that I'm honestly starting to worry about the mental health of those who wrote and approved it...